pkg/queue/cluster_queue_impl.go

```diff
@@ -153,6 +157,7 @@ func (c *clusterQueueBase) Delete(w *kueue.Workload) {
 	key := workload.Key(w)
```

Comment:
🤔 did we forget to lock the mutex? Or is it locked by all callers?

Reply:
can we rename it to `delete` instead?

Reply:
Actually, it's also called from the queue Manager, via the ClusterQueue interface.

But we already access the manager through a mutex. Now I'm wondering why we added a dedicated mutex to the ClusterQueue.

Reply:
Oh it was for the status snapshoter https://github.com/kubernetes-sigs/kueue/pull/1069

Reply:
That's right, sorry for confusion, somehow I missed it in vs-code when using the "all references" option.

IIRC we added the mutex to ClusterQueue so that the `Snapshot` operation (needed for visibility) only requires cq-scoped mutex, and this operation could be expensive.

Reply:
I suppose there might still be problematic scenarios, for example, the event handlers triggering a Workload deletion while the snapshoter is running.

So we should always be locking the mutex. However, maybe we can get rid of the snapshotter in the near future in favor of the visibility API.

Reply:
> I suppose there might still be problematic scenarios, for example, the event handlers triggering a Workload deletion while the snapshoter is running.

The main purpose of the cs-scoped mutex is to make sure the heap structure can be iterated and copied so that we don't miss any element due to concurrent modifications.  

> So we should always be locking the mutex. 

yes

> However, maybe we can get rid of the snapshotter in the near future in favor of the visibility API.

I think we can get rid of the snapshotter that persists data in etcd, but the `Snapshot()` function itself is used by the visibility API to get consistent view of the workloads, so I don't see a way of dropping the cq-scoped mutex, unless:
- we use the manager-scoped mutex, but this would impact performance
- don't use mutex for `Snapshot()` at all, but this would impact consistency of results during updates